opm/autodiff/BlackoilAquiferModel.hpp

```diff
@@ -34,45 +36,61 @@ namespace Opm {
 
         /// Class for handling the blackoil well model.
         template<typename TypeTag>
-        class BlackoilAquiferModel {
-
-        public:
-
-            // ---------      Types      ---------
-            typedef typename GET_PROP_TYPE(TypeTag, ElementContext)      ElementContext;
-            typedef typename GET_PROP_TYPE(TypeTag, Simulator)           Simulator;
-            typedef typename GET_PROP_TYPE(TypeTag, Scalar)              Scalar;
+        class BlackoilAquiferModel : public Ewoms::EclBaseAquiferModel<TypeTag>
```

Comment:
So it will always be a derived class from ebos now?

Reply:
Is this using the "curiously recurring template pattern", or engine concept? It does not look like that to me, as it does not pass the class itself as a template argument. My question is therefore: how is the base class used?

Reply:
well, I don't think that his is the "CRTP", but it is possibly related: it is the same trick as for the well model with #1555, i.e., the interface is defined within ewoms and used directly by `EclProblem`, while the *implementation* can stay in opm-simulators for now. the rationale for these changes is that they allow to "de-frankenstein" the whole thing considerably because such stuff does not have to be bolted on the ebos stuff with ducttape anymore.

> My question is therefore: how is the base class used?

the base class just defines a few member functions that are mandated by the API and which don't do anything. if all of these are defined by the actual implementation, no inheritance is needed. (but this is a bit cumbersome.)

Reply:
> the base class just defines a few member functions that are mandated by the API and which don't do anything.

If this is indeed all, I think inheritance should be dropped, unless there are very many such empty methods (and in that case, a refactoring would probably be due...).

Reply:
I'll add one more thing: if you decide to keep inheritance, it should be documented (in a nearby comment) that it is only done to inherit dummy implementations of methods foo(), bar() etc.

Reply:
> If this is indeed all, I think inheritance should be dropped, unless there are very many such empty methods (and in that case, a refactoring would probably be due...).

done. (It feels a bit strange to be the one who is asked to drop the inheritance design. ;))

@GitPaean: I'll move the class bodies to the `_impl.hpp` file once everything else is settled because changing function signatures would drive me even more crazy than I already am otherwise. I guess that nobody wants that...

Reply:
> because changing function signatures would drive me even more crazy

Compared with other stuff, thinking about styles, readability .. and so on, I really do not mind this kind of `lab` work. Because, I will have a easy to read head file to read later, and easy to tell what is going on with this class. When a class file `.hpp` contains thousands of lines, even there are a lot of junk/unused functions and variables there, I will not be able to tell.